### ISSUE TITLE
Approximate European option price under SABR model 

### ISSUE THREAD

**Issue by cyrilchim** [google/tf-quant-finance#8]
[SABR model](https://en.wikipedia.org/wiki/SABR_volatility_model) has accurate density approximations  (see, e.g., [here](https://arxiv.org/pdf/1107.1834.pdf)). It is of interest to use the approximations to estimate European option prices.

The module implementing this method should live under tf_quant_finance/volatility/sabr_approximation.py. It should support both European puts and calls approximations. Tests should be in sabr_approximation_test.py in the same folder.

**Comment by saxena-ashish-g** [comment]
@joybh98 

Hi Joy, Are you still working on this issue? If you aren't could you please let us know?

-Ashish

**Comment by joybh98** [comment]
@saxena-ashish-g I'm not working on this issue. Whoever wants to take this please go ahead.

**Comment by Rish001** [comment]
@cyrilchim I didnt come across any folder named volatility under tf_quant_finance. Can you please point me to the file?

**Comment by cyrilchim** [comment]
@Rish001 It has been refactored. This has to go to models/sabr/approximations/european_option.py (similarly to [heston structure](https://github.com/google/tf-quant-finance/tree/master/tf_quant_finance/models/heston/approximations))

**Comment by Rish001** [comment]
Do you want this approximation of SABR model to be implemented?


![image](https://user-images.githubusercontent.com/21169674/88480649-c6ff5680-cf74-11ea-8a32-1ba7c75d6b59.png)

**Comment by cyrilchim** [comment]
Yes, Rishav, that is the approximation. One could use it for calibration purposes as described in the paper

**Comment by Rish001** [comment]
What are the possible test cases that my implementation needs to be tested against?

**Comment by cyrilchim** [comment]
Hi Rishav,

You can find different estimated European option prices under the SABR
model here <http://ta.twi.tudelft.nl/mf/users/oosterle/oosterlee/SABRMC.pdf>.
Your approximation should lead to similar results.
Please let me know if you have any further questions. I am happy to help
you debug the code if you get stuck.

Best,
Cyril

**Comment by Rish001** [comment]
How are these integrations working? are these integrations nested in one inside the other?
![image](https://user-images.githubusercontent.com/21169674/89098345-4cb35400-d404-11ea-82f9-d2c1e7338fb2.png)
It would help me if you can please give me a clue how this integral would look like for tenor T2 and beyond

**Comment by cyrilchim** [comment]
@Rish001 The approximation `p(t, ... )` is the forward transition density of the SABR process. Computing option prices could be done by creating a uniform grid of points in `(F, A)` space and computing the average in the formula (2.6) from the original [paper](https://arxiv.org/pdf/1107.1834.pdf). `B` is a zero-coupon bond price which is a standard notion. In case of a constant interest rate `r` it is a discount factor `exp(-r * (T - t))`. (see, e.g., [here](https://www.researchgate.net/profile/Patrick_Hagan3/publication/228463045_Methods_for_constructing_a_yield_curve/links/54db8cda0cf23fe133ad4d01.pdf))

I think the best way to get started is to implement the approximation first. It seems that the original paper is self-contained in terms of notations.

By the time this is implemented, we will have the SABR model in the module and you will be able to use Monte Carlo method to get option prices and get your implementation tested. Does this sound good?

**Comment by Rish001** [comment]
As the mentioned integral is unbounded, I think Monte Carlo integration is to be applied. Do you think any other way to compute the integral?

**Comment by Rish001** [comment]
@cyrilchim I have built a minimal viable code for determining option price for option tenor T1. But i am encountering overflow error which I am unable to remove. I would really appreciate it if you can provide me some helpful direction.
![image](https://user-images.githubusercontent.com/21169674/90872159-d4fc9780-e3b9-11ea-99b3-a896ebf98f3b.png)
Please find the [link ](https://colab.research.google.com/drive/1hC2EkC-k8-qEbxkstteED87yZwFnj5V_?usp=sharing)to code in google colab

**Comment by Rish001** [comment]
Just to leave you an update. I have mostly finished debugging and will be submitting shortly

**Comment by Rish001** [comment]
@cyrilchim could you give me a set of test cases to test my implementation against? I need a number of combinations of alpha, beta,rho, nu,forwards,strikes and time to maturity and corresponding implied volatility values so that I can test my SABR implied vol calculation

**Comment by cyrilchim** [comment]
Hi Rishav,

We do not have explicit tests but you should be able to use current [SABR model](https://github.com/google/tf-quant-finance/blob/master/tf_quant_finance/models/sabr/sabr_model.py) to sample trajectories and estimate European option price (put or call). Your approximation should give similar results
